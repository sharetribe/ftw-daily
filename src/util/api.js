// These helpers are calling FTW's own server-side routes
// so, they are not directly calling Marketplace API or Integration API.
// You can find these api endpoints from 'server/api/...' directory

import { types as sdkTypes, transit } from './sdkLoader';
import config from '../config';
import Decimal from 'decimal.js';

export const apiBaseUrl = () => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the same domain and port as the frontend
  return `${window.location.origin}`;
};

// Application type handlers for JS SDK.
//
// NOTE: keep in sync with `typeHandlers` in `server/api-util/sdk.js`
export const typeHandlers = [
  // Use Decimal type instead of SDK's BigDecimal.
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers, verbose: config.sdk.transitVerbose });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const post = (path, body) => {
  const url = `${apiBaseUrl()}${path}`;
  const options = {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/transit+json',
    },
    body: serialize(body),
  };
  return window.fetch(url, options).then(res => {
    const contentTypeHeader = res.headers.get('Content-Type');
    const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;

    if (res.status >= 400) {
      return res.json().then(data => {
        let e = new Error();
        e = Object.assign(e, data);

        throw e;
      });
    }
    if (contentType === 'application/transit+json') {
      return res.text().then(deserialize);
    } else if (contentType === 'application/json') {
      return res.json();
    }
    return res.text();
  });
};

// Fetch transaction line items from the local API endpoint.
//
// See `server/api/transaction-line-items.js` to see what data should
// be sent in the body.
export const transactionLineItems = body => {
  return post('/api/transaction-line-items', body);
};

// Initiate a privileged transaction.
//
// With privileged transitions, the transactions need to be created
// from the backend. This endpoint enables sending the booking data to
// the local backend, and passing that to the Marketplace API.
//
// See `server/api/initiate-privileged.js` to see what data should be
// sent in the body.
export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};

// Transition a transaction with a privileged transition.
//
// This is similar to the `initiatePrivileged` above. It will use the
// backend for the transition. The backend endpoint will add the
// payment line items to the transition params.
//
// See `server/api/transition-privileged.js` to see what data should
// be sent in the body.
export const transitionPrivileged = body => {
  return post('/api/transition-privileged', body);
};

// Create user with identity provider (e.g. Facebook or Google)
//
// If loginWithIdp api call fails and user can't authenticate to Flex with idp
// we will show option to create a new user with idp.
// For that user needs to confirm data fetched from the idp.
// After the confirmation, this endpoint is called to create a new user with confirmed data.
//
// See `server/api/auth/createUserWithIdp.js` to see what data should
// be sent in the body.
export const createUserWithIdp = body => {
  return post('/api/auth/create-user-with-idp', body);
};
