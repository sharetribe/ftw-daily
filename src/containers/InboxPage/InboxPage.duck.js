import reverse from 'lodash/reverse';
import sortBy from 'lodash/sortBy';
import { storableError } from '../../util/errors';
import { parse } from '../../util/urlHelpers';
import { TRANSITIONS } from '../../util/transaction';
import { addMarketplaceEntities } from '../../ducks/marketplaceData.duck';

const sortedTransactions = txs =>
  reverse(
    sortBy(txs, tx => {
      return tx.attributes ? tx.attributes.lastTransitionedAt : null;
    })
  );

// ================ Action types ================ //

export const FETCH_ORDERS_OR_SALES_REQUEST = 'app/InboxPage/FETCH_ORDERS_OR_SALES_REQUEST';
export const FETCH_ORDERS_OR_SALES_SUCCESS = 'app/InboxPage/FETCH_ORDERS_OR_SALES_SUCCESS';
export const FETCH_ORDERS_OR_SALES_ERROR = 'app/InboxPage/FETCH_ORDERS_OR_SALES_ERROR';

// ================ Reducer ================ //

const entityRefs = entities =>
  entities.map(entity => ({
    id: entity.id,
    type: entity.type,
  }));

const initialState = {
  fetchInProgress: false,
  fetchOrdersOrSalesError: null,
  pagination: null,
  transactionRefs: [],
};

export default function checkoutPageReducer(state = initialState, action = {}) {
  const { type, payload } = action;
  switch (type) {
    case FETCH_ORDERS_OR_SALES_REQUEST:
      return { ...state, fetchInProgress: true, fetchOrdersOrSalesError: null };
    case FETCH_ORDERS_OR_SALES_SUCCESS: {
      const transactions = sortedTransactions(payload.data.data);
      return {
        ...state,
        fetchInProgress: false,
        transactionRefs: entityRefs(transactions),
        pagination: payload.data.meta,
      };
    }
    case FETCH_ORDERS_OR_SALES_ERROR:
      console.error(payload); // eslint-disable-line
      return { ...state, fetchInProgress: false, fetchOrdersOrSalesError: payload };

    default:
      return state;
  }
}

// ================ Action creators ================ //

const fetchOrdersOrSalesRequest = () => ({ type: FETCH_ORDERS_OR_SALES_REQUEST });
const fetchOrdersOrSalesSuccess = response => ({
  type: FETCH_ORDERS_OR_SALES_SUCCESS,
  payload: response,
});
const fetchOrdersOrSalesError = e => ({
  type: FETCH_ORDERS_OR_SALES_ERROR,
  error: true,
  payload: e,
});

// ================ Thunks ================ //

const INBOX_PAGE_SIZE = 10;

export const loadData = (params, search) => (dispatch, getState, sdk) => {
  const { tab } = params;

  const onlyFilterValues = {
    orders: 'order',
    sales: 'sale',
  };

  const onlyFilter = onlyFilterValues[tab];
  if (!onlyFilter) {
    return Promise.reject(new Error(`Invalid tab for InboxPage: ${tab}`));
  }

  dispatch(fetchOrdersOrSalesRequest());

  const { page = 1 } = parse(search);

  const apiQueryParams = {
    only: onlyFilter,
    lastTransitions: TRANSITIONS,
    include: ['provider', 'provider.profileImage', 'customer', 'customer.profileImage', 'booking'],
    'fields.transaction': [
      'lastTransition',
      'lastTransitionedAt',
      'transitions',
      'payinTotal',
      'payoutTotal',
      'listings'
    ],
    'fields.user': ['profile.displayName', 'profile.abbreviatedName'],
    'fields.image': ['variants.square-small', 'variants.square-small2x'],
    page,
    per_page: INBOX_PAGE_SIZE,
  };

  return sdk.transactions
    .query(apiQueryParams)
    .then(response => {
      dispatch(addMarketplaceEntities(response));
      dispatch(fetchOrdersOrSalesSuccess(response));
      return response;
    })
    .catch(e => {
      dispatch(fetchOrdersOrSalesError(storableError(e)));
      throw e;
    });
};
