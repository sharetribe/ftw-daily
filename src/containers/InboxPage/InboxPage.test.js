import React from 'react';
import { renderShallow, renderDeep } from '../../util/test-helpers';
import {
  fakeIntl,
  createCurrentUser,
  createUser,
  createTransaction,
  createBooking,
} from '../../util/test-data';
import { InboxPageComponent, InboxItem, txState } from './InboxPage';
import { TRANSITION_CONFIRM_PAYMENT } from '../../util/transaction';
import { LINE_ITEM_NIGHT } from '../../util/types';

const noop = () => null;

describe('InboxPage', () => {
  it('matches snapshot', () => {
    const provider = createUser('provider-user-id');
    const customer = createUser('customer-user-id');
    const currentUserProvider = createCurrentUser('provider-user-id');
    const currentUserCustomer = createCurrentUser('customer-user-id');

    const startBooking1 = new Date(Date.UTC(2017, 1, 15));
    const endBooking1 = new Date(Date.UTC(2017, 1, 16));

    const booking1 = createBooking('booking1', {
      start: startBooking1,
      end: endBooking1,
      displayStart: startBooking1,
      displayEnd: endBooking1,
    });

    const startBooking2 = new Date(Date.UTC(2017, 2, 15));
    const endBooking2 = new Date(Date.UTC(2017, 2, 16));

    const booking2 = createBooking('booking2', {
      start: startBooking2,
      end: endBooking2,
      displayStart: startBooking2,
      displayEnd: endBooking2,
    });

    const ordersProps = {
      unitType: LINE_ITEM_NIGHT,
      location: { search: '' },
      history: {
        push: () => console.log('HistoryPush called'),
      },
      params: {
        tab: 'orders',
      },
      authInProgress: false,
      currentUser: currentUserProvider,
      currentUserHasListings: false,
      isAuthenticated: false,
      fetchInProgress: false,
      onLogout: noop,
      onManageDisableScrolling: noop,
      transactions: [
        createTransaction({
          id: 'order-1',
          lastTransition: TRANSITION_CONFIRM_PAYMENT,
          customer,
          provider,
          lastTransitionedAt: new Date(Date.UTC(2017, 0, 15)),
          booking: booking1,
        }),
        createTransaction({
          id: 'order-2',
          lastTransition: TRANSITION_CONFIRM_PAYMENT,
          customer,
          provider,
          lastTransitionedAt: new Date(Date.UTC(2016, 0, 15)),
          booking: booking2,
        }),
      ],
      intl: fakeIntl,
      scrollingDisabled: false,
      sendVerificationEmailInProgress: false,
      onResendVerificationEmail: noop,
    };

    const ordersTree = renderShallow(<InboxPageComponent {...ordersProps} />);
    expect(ordersTree).toMatchSnapshot();

    const stateDataOrder = txState(fakeIntl, ordersProps.transactions[0], 'order');

    // Deeply render one InboxItem
    const orderItem = renderDeep(
      <InboxItem
        unitType={LINE_ITEM_NIGHT}
        type="order"
        tx={ordersProps.transactions[0]}
        intl={fakeIntl}
        stateData={stateDataOrder}
      />
    );
    expect(orderItem).toMatchSnapshot();

    const salesProps = {
      unitType: LINE_ITEM_NIGHT,
      location: { search: '' },
      history: {
        push: () => console.log('HistoryPush called'),
      },
      params: {
        tab: 'sales',
      },
      authInProgress: false,
      currentUser: currentUserCustomer,
      currentUserHasListings: false,
      isAuthenticated: false,
      fetchInProgress: false,
      onLogout: noop,
      onManageDisableScrolling: noop,
      transactions: [
        createTransaction({
          id: 'sale-1',
          lastTransition: TRANSITION_CONFIRM_PAYMENT,
          customer,
          provider,
          lastTransitionedAt: new Date(Date.UTC(2017, 0, 15)),
          booking: booking1,
        }),
        createTransaction({
          id: 'sale-2',
          lastTransition: TRANSITION_CONFIRM_PAYMENT,
          customer,
          provider,
          lastTransitionedAt: new Date(Date.UTC(2016, 0, 15)),
          booking: booking2,
        }),
      ],
      intl: fakeIntl,
      scrollingDisabled: false,
      sendVerificationEmailInProgress: false,
      onResendVerificationEmail: noop,
    };

    const salesTree = renderShallow(<InboxPageComponent {...salesProps} />);
    expect(salesTree).toMatchSnapshot();

    const stateDataSale = txState(fakeIntl, salesProps.transactions[0], 'sale');

    // Deeply render one InboxItem
    const saleItem = renderDeep(
      <InboxItem
        unitType={LINE_ITEM_NIGHT}
        type="sale"
        tx={salesProps.transactions[0]}
        intl={fakeIntl}
        stateData={stateDataSale}
      />
    );
    expect(saleItem).toMatchSnapshot();
  });
});
