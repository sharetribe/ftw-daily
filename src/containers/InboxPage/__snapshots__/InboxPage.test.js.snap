// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InboxPage matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="InboxPage.ordersTitle"
>
  <LayoutSideNavigation
    className={null}
    containerClassName={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent))
        className="topbar"
        currentPage="InboxPage"
        desktopClassName="desktopTopbar"
        mobileRootClassName="mobileTopbar"
      />
    </LayoutWrapperTopbar>
    <LayoutWrapperSideNav
      className="navigation"
      rootClassName={null}
      tabs={null}
    >
      <h1
        className="title"
      >
        <MemoizedFormattedMessage
          id="InboxPage.title"
        />
      </h1>
      <TabNav
        className={null}
        rootClassName="tabs"
        tabClassName={null}
        tabRootClassName="tab"
        tabs={
          Array [
            Object {
              "linkProps": Object {
                "name": "InboxPage",
                "params": Object {
                  "tab": "orders",
                },
              },
              "selected": true,
              "text": <span>
                <Memo(MemoizedFormattedMessage)
                  id="InboxPage.ordersTabTitle"
                />
              </span>,
            },
            Object {
              "linkProps": Object {
                "name": "InboxPage",
                "params": Object {
                  "tab": "sales",
                },
              },
              "selected": false,
              "text": <span>
                <Memo(MemoizedFormattedMessage)
                  id="InboxPage.salesTabTitle"
                />
              </span>,
            },
          ]
        }
      />
    </LayoutWrapperSideNav>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <ul
        className="itemList"
      >
        <li
          className="listItem"
          key="order-1"
        >
          <InboxItem
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            stateData={
              Object {
                "bookingClassName": "bookingNoActionNeeded",
                "lastTransitionedAtClassName": "lastTransitionedAtEmphasized",
                "nameClassName": "nameNotEmphasized",
                "state": "InboxPage.stateRequested",
                "stateClassName": "stateActionNeeded",
              }
            }
            tx={
              Object {
                "attributes": Object {
                  "createdAt": 2017-05-01T00:00:00.000Z,
                  "lastTransition": "transition/confirm-payment",
                  "lastTransitionedAt": 2017-01-15T00:00:00.000Z,
                  "lineItems": Array [
                    Object {
                      "code": "line-item/night",
                      "includeFor": Array [
                        "customer",
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                      "quantity": "1",
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                    },
                    Object {
                      "code": "line-item/provider-commission",
                      "includeFor": Array [
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                    },
                  ],
                  "payinTotal": Money {
                    "_sdkType": "Money",
                    "amount": 1000,
                    "currency": "USD",
                  },
                  "payoutTotal": Money {
                    "_sdkType": "Money",
                    "amount": 900,
                    "currency": "USD",
                  },
                  "transitions": Array [
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:00.000Z,
                      "transition": "transition/request-payment",
                    },
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:01.000Z,
                      "transition": "transition/confirm-payment",
                    },
                    Object {
                      "by": "provider",
                      "createdAt": 2017-06-01T00:00:00.000Z,
                      "transition": "transition/accept",
                    },
                  ],
                },
                "booking": Object {
                  "attributes": Object {
                    "displayEnd": 2017-02-16T00:00:00.000Z,
                    "displayStart": 2017-02-15T00:00:00.000Z,
                    "end": 2017-02-16T00:00:00.000Z,
                    "start": 2017-02-15T00:00:00.000Z,
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "booking1",
                  },
                  "type": "booking",
                },
                "customer": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "customer-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "customer-user-id",
                  },
                  "type": "user",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "order-1",
                },
                "listing": null,
                "provider": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "provider-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "provider-user-id",
                  },
                  "type": "user",
                },
                "reviews": Array [],
                "type": "transaction",
              }
            }
            type="order"
            unitType="line-item/night"
          />
        </li>
        <li
          className="listItem"
          key="order-2"
        >
          <InboxItem
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            stateData={
              Object {
                "bookingClassName": "bookingNoActionNeeded",
                "lastTransitionedAtClassName": "lastTransitionedAtEmphasized",
                "nameClassName": "nameNotEmphasized",
                "state": "InboxPage.stateRequested",
                "stateClassName": "stateActionNeeded",
              }
            }
            tx={
              Object {
                "attributes": Object {
                  "createdAt": 2017-05-01T00:00:00.000Z,
                  "lastTransition": "transition/confirm-payment",
                  "lastTransitionedAt": 2016-01-15T00:00:00.000Z,
                  "lineItems": Array [
                    Object {
                      "code": "line-item/night",
                      "includeFor": Array [
                        "customer",
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                      "quantity": "1",
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                    },
                    Object {
                      "code": "line-item/provider-commission",
                      "includeFor": Array [
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                    },
                  ],
                  "payinTotal": Money {
                    "_sdkType": "Money",
                    "amount": 1000,
                    "currency": "USD",
                  },
                  "payoutTotal": Money {
                    "_sdkType": "Money",
                    "amount": 900,
                    "currency": "USD",
                  },
                  "transitions": Array [
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:00.000Z,
                      "transition": "transition/request-payment",
                    },
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:01.000Z,
                      "transition": "transition/confirm-payment",
                    },
                    Object {
                      "by": "provider",
                      "createdAt": 2017-06-01T00:00:00.000Z,
                      "transition": "transition/accept",
                    },
                  ],
                },
                "booking": Object {
                  "attributes": Object {
                    "displayEnd": 2017-03-16T00:00:00.000Z,
                    "displayStart": 2017-03-15T00:00:00.000Z,
                    "end": 2017-03-16T00:00:00.000Z,
                    "start": 2017-03-15T00:00:00.000Z,
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "booking2",
                  },
                  "type": "booking",
                },
                "customer": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "customer-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "customer-user-id",
                  },
                  "type": "user",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "order-2",
                },
                "listing": null,
                "provider": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "provider-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "provider-user-id",
                  },
                  "type": "user",
                },
                "reviews": Array [],
                "type": "transaction",
              }
            }
            type="order"
            unitType="line-item/night"
          />
        </li>
      </ul>
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className={null}
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSideNavigation>
</Page>
`;

exports[`InboxPage matches snapshot 2`] = `
<div
  className="item"
>
  <div
    className="itemAvatar"
  >
    <a
      className="root"
      href="/u/provider-user-id"
      onClick={[Function]}
      onMouseOver={[Function]}
      onTouchStart={[Function]}
      style={Object {}}
      title="provider-user-id display name"
    >
      <span
        className="initials"
      >
        TT
      </span>
    </a>
  </div>
  <a
    className="itemLink"
    href="/order/order-1/details"
    onClick={[Function]}
    onMouseOver={[Function]}
    onTouchStart={[Function]}
    style={Object {}}
    title={null}
  >
    <div
      className="rowNotificationDot"
    />
    <div
      className="itemInfo"
    >
      <div
        className="itemUsername nameNotEmphasized"
      >
        <span
          className=""
        >
          provider-user-id display name
        </span>
      </div>
      <div
        className="bookingInfoWrapper bookingNoActionNeeded"
      >
        <div
          className="bookingInfo bookingNoActionNeeded"
        >
          <span
            className="dateSection"
          >
            Feb 15 -
          </span>
          <span
            className="dateSection"
          >
            Feb 16
          </span>
        </div>
      </div>
    </div>
    <div
      className="itemState"
    >
      <div
        className="stateName stateActionNeeded"
      >
        InboxPage.stateRequested
      </div>
      <div
        className="lastTransitionedAt lastTransitionedAtEmphasized"
        title="2017-01-15 00:00"
      >
        2017-01-15
      </div>
    </div>
  </a>
</div>
`;

exports[`InboxPage matches snapshot 3`] = `
<Page
  scrollingDisabled={false}
  title="InboxPage.salesTitle"
>
  <LayoutSideNavigation
    className={null}
    containerClassName={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent))
        className="topbar"
        currentPage="InboxPage"
        desktopClassName="desktopTopbar"
        mobileRootClassName="mobileTopbar"
      />
    </LayoutWrapperTopbar>
    <LayoutWrapperSideNav
      className="navigation"
      rootClassName={null}
      tabs={null}
    >
      <h1
        className="title"
      >
        <MemoizedFormattedMessage
          id="InboxPage.title"
        />
      </h1>
      <TabNav
        className={null}
        rootClassName="tabs"
        tabClassName={null}
        tabRootClassName="tab"
        tabs={
          Array [
            Object {
              "linkProps": Object {
                "name": "InboxPage",
                "params": Object {
                  "tab": "orders",
                },
              },
              "selected": false,
              "text": <span>
                <Memo(MemoizedFormattedMessage)
                  id="InboxPage.ordersTabTitle"
                />
              </span>,
            },
            Object {
              "linkProps": Object {
                "name": "InboxPage",
                "params": Object {
                  "tab": "sales",
                },
              },
              "selected": true,
              "text": <span>
                <Memo(MemoizedFormattedMessage)
                  id="InboxPage.salesTabTitle"
                />
              </span>,
            },
          ]
        }
      />
    </LayoutWrapperSideNav>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <ul
        className="itemList"
      >
        <li
          className="listItem"
          key="sale-1"
        >
          <InboxItem
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            stateData={
              Object {
                "bookingClassName": "bookingActionNeeded",
                "lastTransitionedAtClassName": "lastTransitionedAtEmphasized",
                "nameClassName": "nameEmphasized",
                "state": "InboxPage.statePending",
                "stateClassName": "stateActionNeeded",
              }
            }
            tx={
              Object {
                "attributes": Object {
                  "createdAt": 2017-05-01T00:00:00.000Z,
                  "lastTransition": "transition/confirm-payment",
                  "lastTransitionedAt": 2017-01-15T00:00:00.000Z,
                  "lineItems": Array [
                    Object {
                      "code": "line-item/night",
                      "includeFor": Array [
                        "customer",
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                      "quantity": "1",
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                    },
                    Object {
                      "code": "line-item/provider-commission",
                      "includeFor": Array [
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                    },
                  ],
                  "payinTotal": Money {
                    "_sdkType": "Money",
                    "amount": 1000,
                    "currency": "USD",
                  },
                  "payoutTotal": Money {
                    "_sdkType": "Money",
                    "amount": 900,
                    "currency": "USD",
                  },
                  "transitions": Array [
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:00.000Z,
                      "transition": "transition/request-payment",
                    },
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:01.000Z,
                      "transition": "transition/confirm-payment",
                    },
                    Object {
                      "by": "provider",
                      "createdAt": 2017-06-01T00:00:00.000Z,
                      "transition": "transition/accept",
                    },
                  ],
                },
                "booking": Object {
                  "attributes": Object {
                    "displayEnd": 2017-02-16T00:00:00.000Z,
                    "displayStart": 2017-02-15T00:00:00.000Z,
                    "end": 2017-02-16T00:00:00.000Z,
                    "start": 2017-02-15T00:00:00.000Z,
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "booking1",
                  },
                  "type": "booking",
                },
                "customer": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "customer-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "customer-user-id",
                  },
                  "type": "user",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "sale-1",
                },
                "listing": null,
                "provider": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "provider-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "provider-user-id",
                  },
                  "type": "user",
                },
                "reviews": Array [],
                "type": "transaction",
              }
            }
            type="sale"
            unitType="line-item/night"
          />
        </li>
        <li
          className="listItem"
          key="sale-2"
        >
          <InboxItem
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            stateData={
              Object {
                "bookingClassName": "bookingActionNeeded",
                "lastTransitionedAtClassName": "lastTransitionedAtEmphasized",
                "nameClassName": "nameEmphasized",
                "state": "InboxPage.statePending",
                "stateClassName": "stateActionNeeded",
              }
            }
            tx={
              Object {
                "attributes": Object {
                  "createdAt": 2017-05-01T00:00:00.000Z,
                  "lastTransition": "transition/confirm-payment",
                  "lastTransitionedAt": 2016-01-15T00:00:00.000Z,
                  "lineItems": Array [
                    Object {
                      "code": "line-item/night",
                      "includeFor": Array [
                        "customer",
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                      "quantity": "1",
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": 1000,
                        "currency": "USD",
                      },
                    },
                    Object {
                      "code": "line-item/provider-commission",
                      "includeFor": Array [
                        "provider",
                      ],
                      "lineTotal": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                      "reversal": false,
                      "unitPrice": Money {
                        "_sdkType": "Money",
                        "amount": -100,
                        "currency": "USD",
                      },
                    },
                  ],
                  "payinTotal": Money {
                    "_sdkType": "Money",
                    "amount": 1000,
                    "currency": "USD",
                  },
                  "payoutTotal": Money {
                    "_sdkType": "Money",
                    "amount": 900,
                    "currency": "USD",
                  },
                  "transitions": Array [
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:00.000Z,
                      "transition": "transition/request-payment",
                    },
                    Object {
                      "by": "customer",
                      "createdAt": 2017-05-01T00:00:01.000Z,
                      "transition": "transition/confirm-payment",
                    },
                    Object {
                      "by": "provider",
                      "createdAt": 2017-06-01T00:00:00.000Z,
                      "transition": "transition/accept",
                    },
                  ],
                },
                "booking": Object {
                  "attributes": Object {
                    "displayEnd": 2017-03-16T00:00:00.000Z,
                    "displayStart": 2017-03-15T00:00:00.000Z,
                    "end": 2017-03-16T00:00:00.000Z,
                    "start": 2017-03-15T00:00:00.000Z,
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "booking2",
                  },
                  "type": "booking",
                },
                "customer": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "customer-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "customer-user-id",
                  },
                  "type": "user",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "sale-2",
                },
                "listing": null,
                "provider": Object {
                  "attributes": Object {
                    "banned": false,
                    "deleted": false,
                    "profile": Object {
                      "abbreviatedName": "TT",
                      "displayName": "provider-user-id display name",
                    },
                  },
                  "id": UUID {
                    "_sdkType": "UUID",
                    "uuid": "provider-user-id",
                  },
                  "type": "user",
                },
                "reviews": Array [],
                "type": "transaction",
              }
            }
            type="sale"
            unitType="line-item/night"
          />
        </li>
      </ul>
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className={null}
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSideNavigation>
</Page>
`;

exports[`InboxPage matches snapshot 4`] = `
<div
  className="item"
>
  <div
    className="itemAvatar"
  >
    <a
      className="root"
      href="/u/customer-user-id"
      onClick={[Function]}
      onMouseOver={[Function]}
      onTouchStart={[Function]}
      style={Object {}}
      title="customer-user-id display name"
    >
      <span
        className="initials"
      >
        TT
      </span>
    </a>
  </div>
  <a
    className="itemLink"
    href="/sale/sale-1/details"
    onClick={[Function]}
    onMouseOver={[Function]}
    onTouchStart={[Function]}
    style={Object {}}
    title={null}
  >
    <div
      className="rowNotificationDot"
    >
      <div
        className="notificationDot"
      />
    </div>
    <div
      className="itemInfo"
    >
      <div
        className="itemUsername nameEmphasized"
      >
        <span
          className=""
        >
          customer-user-id display name
        </span>
      </div>
      <div
        className="bookingInfoWrapper bookingActionNeeded"
      >
        <div
          className="bookingInfo bookingActionNeeded"
        >
          <span
            className="dateSection"
          >
            Feb 15 -
          </span>
          <span
            className="dateSection"
          >
            Feb 16
          </span>
        </div>
      </div>
    </div>
    <div
      className="itemState"
    >
      <div
        className="stateName stateActionNeeded"
      >
        InboxPage.statePending
      </div>
      <div
        className="lastTransitionedAt lastTransitionedAtEmphasized"
        title="2017-01-15 00:00"
      >
        2017-01-15
      </div>
    </div>
  </a>
</div>
`;
